### ISSUE TITLE
[Script Recorder] Notarise Electron App

### ISSUE THREAD

**Issue by drewpost** [elastic/synthetics-recorder#51]
On macOS, we need to notarise our application to prevent Gatekeeper from throwing an error when it's run that requires a user to process a security exception. This requires admin access which a lot of users won't have, is a frustrating step for users who do have permissions and is reputation ally damaging.

https://kilianvalkhof.com/2019/electron/notarizing-your-electron-application/

**Comment by brunerd** [comment]
Hi, 
I'm the macOS Endpoint engineer for our Elastic fleet and I've also have admin duties on our Apple Dev account, this issue came to me via [Salesforce ticket 00829187](https://elastic.my.salesforce.com/5004M00000lXPyx).

Currently we've got App IDs for a few Elastic products:
![image](https://user-images.githubusercontent.com/9872375/145462962-7ab0362e-5211-49c9-b63f-7ce8b7aca4bf.png)

The Endgame and Agent folks are the main active ones, I usually do renewal on the certs and profiles, or approved new devices. You may only need a Developer ID to sign your self-distributed apps, but I'm not sure of what the entitlements that your Synthetics recorder might be using? Maybe none?

Otherwise it looks like we'd need to:
* Add both @vigneshshanmugam  and @justinkambic to the Dev program
* Create a new App ID (e.g. co.elastic.syntheticsthingamabob)
* Create a "Developer ID Application" certificate to sign the app (if there is a pkg installer that is a separate cert)

I am might be missing some things since I admin this but am not a dev using these services day-today. So indeed you might benefit from asking the other teams at Elastic who are releasing about this. Once we've got a good idea of App ID names, any entitlements or Services used, etc. we can move forward. Also just checking since this app is already being used in production, InfoSec has already had a chance to do whatever reviews they need to?

**Comment by justinkambic** [comment]
@brunerd looking at the list of entitlements, at a glance I don't think we need any of them per se.

> Also just checking since this app is already being used in production, InfoSec has already had a chance to do whatever reviews they need to?

The app is published as a dev preview at the moment, it's not seeing actual production use yet. We'll start the process with InfoSec soon though. I think your assessment is correct.

> Add both @vigneshshanmugam and @justinkambic to the Dev program

Could we add a group instead of individuals?

> Create a new App ID (e.g. co.elastic.syntheticsthingamabob)

This should be `co.elastic.synthetics-recorder` or `co.elastic.syntheticsRecorder` AFAIK. cc @drewpost

**Comment by justinkambic** [comment]
As an aside, this is the reason we want to get progress on this issue. Users are greeted with this if they try to open from the Applications menu:

![image](https://user-images.githubusercontent.com/18429259/146982736-1255110a-9039-4719-b216-7586912fd108.png)

If they do the "command+click" method of opening it, they're still given a scary warning message:

![image](https://user-images.githubusercontent.com/18429259/146982983-939f99bb-9d93-4955-9cdd-89329fb23403.png)

**Comment by brunerd** [comment]
Apple Developer accounts are created per user in "App Store Connect" and it can't leverage our Okta groups, it doesn't have that kind of integration.

Requests come in via the it-eng github, here's some examples of the format and types of requests:
Add new employee to Apple Developer Account - https://github.com/elastic/it-eng/issues/850
Add new Endpoint developer VM UUID to Apple Provisioning Profiles - https://github.com/elastic/it-eng/issues/839

We'll also need to set up the App ID (the  bundle ID of the app), the Provisioning Profiles and assign tose profiles to the UUIDs of the Macs being used for dev work.

**Comment by justinkambic** [comment]
Update here; we've created an issue to track the addition of users and provisioning of an app ID: (private repo) https://github.com/elastic/it-eng/issues/1146.

We need some additional information, namely the UUID of the other developers' laptops not included in that list, as well as a final decision on the App ID to use when signing the bundle.

When we're ready we can let the IT team know and they'll set up and provision the necessary resources for us. The rest of the team is out until the new year so we will need to wait a bit before we can complete this setup process. Once we're past that we can work to notarize the application and issue a new release that won't issue the warning.

**Comment by brunerd** [comment]
OK, I've created both the App ID `co.elastic.synthetics-recorder` and a "Developer ID Application" Provisioning Profile named "Synthetics Recorder" associated the certificate `Elasticsearch, Inc (Developer ID Application), Jun 02, 2023` (there are 3 in the portal, I chose the date furthest out). Looking closer at the way the Elastic Agent folks have things set up, the UUID assignment _only_ comes into play for **Development** Profiles. Since this is a "Developer ID Application" it isn't even capable of assignment like that, so we indeed should not need this. Although _already_ added the UUIDs for @lucasfcosta  @justinkambic but @vigneshshanmugam I didn't add yours, so that'll be a fine test, it shouldn't be needed for this type of Profile. All the invites are away just now, so look out for those, accept them and sign into XCode and fiddle around 🎻 let me know if you need additional administrative fiddling.

Also, the `allow-unsigned-executable-memory` entitlement seems to only be a local setting you need to set, it's not something done at the Developer portal level.

**Comment by justinkambic** [comment]
Thanks @brunerd - I will try to let you know today if I encounter any snags.

**Comment by drewpost** [comment]
Hey @justinkambic - any update on this?

**Comment by justinkambic** [comment]
Hi @drewpost - yes we are progressing on this. Notarization is a late-stage milestone of https://github.com/elastic/observability-robots/issues/1113, which is tracking the overall automation of all the release stages we'll need in order to be GA.

We've recently merged https://github.com/elastic/synthetics-recorder/pull/238, which takes care of signing the binaries with Elastic's official developer certificate. We will be doing some experimentation around notarizing these binaries after we ship our next release, which I'm hoping will be this week sometime.

Once we have shipped a signed binary, I'll experiment with Apple's notarization service to see what additional work is needed (specifying entitlements, is it the right kind of cert, etc.), and we'll make whatever requisite changes to the CI process as needed. At that point, we should be able to automatically trigger a release that will take care of the signing, communication with the Apple service, and publishing the final release binary. It is possible there's a significant amount of discrete scope in that last bit that we won't uncover until we're working on it.

**Comment by justinkambic** [comment]
After some additional investigation, you cannot submit to Apple's notarization service without access to the keychain that contains the certificate used for signing, so we're going to have to develop this piece of the pipeline completely within our CI environment. cc @v1v

**Comment by v1v** [comment]
> After some additional investigation, you cannot submit to Apple's notarization service without access to the keychain that contains the certificate used for signing, so we're going to have to develop this piece of the pipeline completely within our CI environment. cc @v1v

As long as the the notorization runs in the same CI worker where the `electron-build` was executed previously, then it should be ok, right? If that's the case, then there is only need to build the notarization script that consumes the list of signed artifacts to be submitted to the Apple's notorization.

Let me ping @mgreau , since he might have some ideas about this.

The original support to sign the apple binaries was done in -> https://github.com/elastic/synthetics-recorder/pull/238

**Comment by kyungeunni** [comment]
I've started creating a draft adding notarization script to afterSign hook: https://github.com/elastic/synthetics-recorder/pull/290
I've pushed the custom tag and triggered the release pipeline, but it failed because the worker has xcode 12 or earlier installed so it can not use the latest tool, [notarytool](https://developer.apple.com/videos/play/wwdc2021/10261/). 
Even though we decide to use the legacy tool, we are missing some credentials in the keychain, so here are two options we can take:

### option 1. Using Worker having xcode13 installed:
- We need the profile in the keychain mentioned in this [comment](https://github.com/elastic/synthetics-recorder/issues/51#issuecomment-1005257998) (Is this the same item in the keychain entry as we use for signing? if yes, we don't need to add anything)

### option 2. Using current Worker:
- We need to provide `appleId` and  `appleIdPassword` mentioned in the Legacy option: https://github.com/electron/electron-notarize#method-notarizeopts-promisevoid
- Also need to follow safety recommendations when using `appleIdPassword`: https://github.com/electron/electron-notarize#safety-when-using-appleidpassword)


@v1v Could you advise which option would be possible to take for us going forward?

**Comment by v1v** [comment]
> @v1v Could you advise which option would be possible to take for us going forward?

We might need to involve the release and CI systems team as they are the system owners and might know further about what certificates or xcode are available in the existing workers. @elastic/release-eng and @elastic/ci-systems , would you mind helping on this? @kyungeunni asked a few questions in https://github.com/elastic/synthetics-recorder/issues/51#issuecomment-1226033007 that you probably can provide further insights.

**Comment by nkammah** [comment]
We do have a [task](https://github.com/elastic/release-eng/issues/92) to upgrade the macOS signing workers to a newer OS, which would include upgrading Xcode. From that angle, I can see going down the [option1](https://github.com/elastic/synthetics-recorder/issues/51#issuecomment-1226033007). 

With that said, the work has been put on hold so far due to: 
* how short staff the Release Eng team is
* how critical the 2 signing macs are within the release workflow

@kyungeunni what is your rough timeline for wrapping this up ? Would you potentially have some bandwidth in helping to manually provision ([instructions](https://github.com/elastic/infra/issues/15804#issuecomment-635521954)) and validate the macOS worker if that's an option - the Release Eng team would normally do it, but as mentioned above, we have no bandwidth at the moment, and a collaboration may be the most effective approach. 


cc @mgreau

**Comment by kyungeunni** [comment]
Hello @nkammah! 
the aim is to get it done before 8.5 release date. I'd love to try manually provisioning it with the given instruction. Thanks for the pointer!

**Comment by nkammah** [comment]
@kyungeunni I re-imaged `worker-c07yc0cejyvy` and it now has Xcode 13.4.1 and

```
bash-3.2$ xcrun notarytool
OVERVIEW: Manage submissions to the Apple notary service

USAGE: notarytool <subcommand>

OPTIONS:
  --version               Show the version.
  -h, --help              Show help information.

SUBCOMMANDS:
  store-credentials       Save Notary service credentials to the Keychain. Leave credential options not specified for interactive prompts.
  submit                  Submit an archive to the Notary service
  info                    Get status information for a submission
  wait                    Wait for completion of a previous submission
  history                 Get a list of previous submissions for your team
  log                     Retrieve notarization log for a single completed submission

  See 'notarytool help <subcommand>' for detailed help.
```

I am validating that worker onto https://internal-ci.elastic.co/. Do you have a Jenkins job we can use to validate that it works on this worker for your app?

**Comment by kyungeunni** [comment]
Hi @nkammah, thanks for this update!! 🥇 

I've triggered a [Jenkins job](https://internal-ci.elastic.co/blue/organizations/jenkins/elastic%2Bsynthetics-recorder%2Brelease/detail/v0.0.1-notarization.2/1/pipeline/) that uses `notarytool` and seems like it works! (The job failed due to a credential misconfiguration, not because of the worker)

**Comment by nkammah** [comment]
> The job failed due to a credential misconfiguration

Is this something to be fixed on your end or on the worker's end?

**Comment by nkammah** [comment]
Our Release Eng. team signing expert is not available atm, but it seems like [this is](https://github.com/elastic/infra/blob/master/cd/release/endpoint-macos-artifact-signer/ci.sh#L83) how we do signing of other apps - would a similar approach work for your app?

**Comment by kyungeunni** [comment]
Thanks for sharing the reference! 👍 I'm trying out Electron focused notarisation tooling (https://github.com/electron/electron-notarize) so once we have the proper credential available for the recorder(I can see Endgame team also uses [app-specific password](https://github.com/elastic/infra/blob/master/cd/release/endpoint-macos-artifact-signer/ci.sh#L73-L74)), I think that should be an excellent first step.

Also, I noticed we'd need to review the entitlements list too. It'd be great to consult with the expert in Realease Eng. team! (ofc when they are available :) )

**Comment by kyungeunni** [comment]
Hi @v1v

Is it possible for us to use [notarize-macos-artifacts](https://internal-ci.elastic.co/job/elastic+unified-release+master+notarize-macos-artifacts/) job for notarising? I found out we use the same credentials for all the mac apps. The [scripts](https://github.com/elastic/infra/tree/master/cd/release/elastic-macos-notarizer) seems to be quite general so I was wondering if it'd be possible and also make sense to reuse the existing script. (Also I'd like to avoid exporting credentials to be used in the release scripts as we had security concerns)

Here you can find threads and references regarding the notarisation: https://github.com/elastic/it-eng/issues/1654#issuecomment-1247620277

**Comment by nkammah** [comment]
@kyungeunni @v1v feel free to take a stab at using the existing [Jenkins job](https://internal-ci.elastic.co/job/elastic+unified-release+master+notarize-macos-single-package/build?delay=0sec) and reporting what you find out

**Comment by kyungeunni** [comment]
@nkammah Thanks for the link 🎉 I've tried it, but it seems to fail when copying files around 🤔  [Here's](https://internal-ci.elastic.co/job/elastic+unified-release+master+notarize-macos-single-package/23/console) the latest test I've done.

**Comment by nkammah** [comment]
@kyungeunni I believe the problem is not with failing to copy the files, but rather than the app, or some portion of it, was already signed. We do have exceptions on some files already signed (see [here](https://github.com/elastic/infra/blob/master/cd/release/elastic-macos-notarizer/sign.sh#L171-L174)) but it seems like in this case, this one is not listed.
<img width="2037" alt="image" src="https://user-images.githubusercontent.com/124953/191215007-7703777b-155b-4963-aefa-39f184aefaaa.png">

Did you sign `'./elastic-synthetics-recorder-notarize-2.zip.extracted/Elastic Synthetics Recorder.app/Contents/Frameworks/Elastic Synthetics Recorder Helper.app'.` before submitting it to the Jenkins job? If so, can you try again with an un-signed version?

**Comment by kyungeunni** [comment]
Actually, I uploaded un-signed one(the one I built in my local machine). It is odd that it says it's already signed 🤔 .

I thought it was something related to the artifacts not found in the expected directory as I was seeing these error code:
![image](https://user-images.githubusercontent.com/16660866/191229756-a8dbbe5b-6180-4229-99bb-896735b8fb3b.png)
![image](https://user-images.githubusercontent.com/16660866/191229790-bc8ec78a-9c4e-453f-9074-4cfa07241fd8.png)

**Comment by nkammah** [comment]
The signing script starts by detecting all the path within the .zip file to be signed - in your app's case, it looks as below:
```
Elastic Synthetics Recorder.app/Contents/Resources/app.asar.unpacked/node_modules/sharp/vendor/8.12.2/darwin-x64/lib/libvips-cpp.42.dylib
Elastic Synthetics Recorder.app/Contents/Resources/app.asar.unpacked/node_modules/sharp/vendor/8.12.2/darwin-arm64v8/lib/libvips-cpp.42.dylib
Elastic Synthetics Recorder.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework
Elastic Synthetics Recorder.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Libraries/libEGL.dylib
Elastic Synthetics Recorder.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Libraries/libvk_swiftshader.dylib
Elastic Synthetics Recorder.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Libraries/libGLESv2.dylib
Elastic Synthetics Recorder.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Libraries/libffmpeg.dylib
Elastic Synthetics Recorder.app/Contents/Frameworks/ReactiveObjC.framework/Versions/A/ReactiveObjC
Elastic Synthetics Recorder.app/Contents/Frameworks/Squirrel.framework/Versions/A/Squirrel
Elastic Synthetics Recorder.app/Contents/Frameworks/Mantle.framework/Versions/A/Mantle
Elastic Synthetics Recorder.app
Elastic Synthetics Recorder.app/Contents/Frameworks/Elastic Synthetics Recorder Helper.app
Elastic Synthetics Recorder.app/Contents/Frameworks/Elastic Synthetics Recorder Helper (Plugin).app
Elastic Synthetics Recorder.app/Contents/Frameworks/Elastic Synthetics Recorder Helper (GPU).app
Elastic Synthetics Recorder.app/Contents/Frameworks/Elastic Synthetics Recorder Helper (Renderer).app
```

The error we were receiving was that `./elastic-synthetics-recorder-notarize-2.zip.extracted/Elastic Synthetics Recorder.app/Contents/Frameworks/Elastic Synthetics Recorder Helper.app` was already signed. 

Looking at the order of path to sign, it seems like we're signing, in that order

* Elastic Synthetics Recorder.app
* Elastic Synthetics Recorder.app/Contents/Frameworks/Elastic Synthetics Recorder Helper.app

Given that the command sign options we use are as below
```
--deep --force --timestamp --options runtime --strict --verbose=9
```

my initial theory was that the `--deep` argument was causing a recursive signing, and that indeed `Elastic Synthetics Recorder Helper.app` was already signed once when we sign `Elastic Synthetics Recorder.app`.

For experimentation sake, I removed the `--deep` flag locally, and re-attempted the signing. It failed with a stapling error:

```
Processing: /private/var/lib/jenkins/workspace/elastic+unified-release+master+notarize-macos-single-package/cd/release/elastic-macos-notarizer/workspace/extracted-artifacts/synthetics-recorder-0.0.1-notarization.3-mac-arm64.zip.extracted/Elastic Synthetics Recorder.app
CloudKit query for Elastic Synthetics Recorder.app (2/d64416bdc2248c4768a1961aed70afd6d873298d) failed due to "Record not found".
Could not find base64 encoded ticket in response for 2/d64416bdc2248c4768a1961aed70afd6d873298d
The staple and validate action failed! Error 65.
```

I don't know if this is because I tried to sign an arm64 binary on a non-arm mac.

**Comment by nkammah** [comment]
Next step - try the [other](https://github.com/elastic/infra/tree/master/cd/release/endpoint-macos-artifact-signer) notarization script provided by endgame.

**Comment by nkammah** [comment]
@kyungeunni I've been going on and on, trying different things, but it seems like the stapling is failing and I suspect because of an issue with the code structure. 

```
[notarizer] Some notarizations completed with errors.
```

It links to a more detailed log visible [here](https://osxapps-ssl.itunes.apple.com/itunes-assets/Enigma112/v4/19/d7/2b/19d72ba9-1b61-1771-57a8-8cce5e9d3d33/developer_log.json?accessKey=1664048613_6198925411980496636_azmFOBEM5dOhqFewo55gh4410VfeSIf9gZaNTL7ScJlxFCEG1KQNcFI4OXoUFqkAHjlWl4k3Nhk%2BTdBgE49rAAJPST04KILmMOAfC0TVTbhYJ23NLo2ehCSOXhMT8IB4gVlmAHMn1Q4nd21RA%2F968k1N%2FCH0E82CBsmg%2B4%2FCFwk%3D)

```
{
  "logFormatVersion": 1,
  "jobId": "a9ac778e-a9e1-4510-862a-d95b6256763d",
  "status": "Invalid",
  "statusSummary": "Archive contains critical validation errors",
  "statusCode": 4000,
  "archiveFilename": "synthetics-recorder-0.0.1-beta.6-mac-arm64.zip.zip",
  "uploadDate": "2022-09-22T13:37:17Z",
  "sha256": "16537c39c6a2de6ed437d384077ae1506fe85f54947b031f90b4ebfb9d5d9d5d",
  "ticketContents": null,
  "issues": [
    {
      "severity": "error",
      "code": null,
      "path": "synthetics-recorder-0.0.1-beta.6-mac-arm64.zip.zip/workspace/extracted-artifacts/synthetics-recorder-0.0.1-beta.6-mac-arm64.zip.extracted/Elastic Synthetics Recorder.app/Contents/MacOS/Elastic Synthetics Recorder",
      "message": "The signature of the binary is invalid.",
      "docUrl": null,
      "architecture": "arm64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "synthetics-recorder-0.0.1-beta.6-mac-arm64.zip.zip/workspace/extracted-artifacts/synthetics-recorder-0.0.1-beta.6-mac-arm64.zip.extracted/Elastic Synthetics Recorder.app/Contents/Resources/local-browsers/chromium-978106/chrome-mac/Chromium.app/Contents/MacOS/Chromium",
      "message": "The signature of the binary is invalid.",
      "docUrl": null,
      "architecture": "arm64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "synthetics-recorder-0.0.1-beta.6-mac-arm64.zip.zip/workspace/extracted-artifacts/synthetics-recorder-0.0.1-beta.6-mac-arm64.zip.extracted/Elastic Synthetics Recorder.app/Contents/Resources/local-browsers/chromium-978106/chrome-mac/Chromium.app/Contents/Frameworks/Chromium Framework.framework/Chromium Framework",
      "message": "The signature of the binary is invalid.",
      "docUrl": null,
      "architecture": "arm64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "synthetics-recorder-0.0.1-beta.6-mac-arm64.zip.zip/workspace/extracted-artifacts/synthetics-recorder-0.0.1-beta.6-mac-arm64.zip.extracted/Elastic Synthetics Recorder.app/Contents/Frameworks/Electron Framework.framework/Electron Framework",
      "message": "The signature of the binary is invalid.",
      "docUrl": null,
      "architecture": "arm64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "synthetics-recorder-0.0.1-beta.6-mac-arm64.zip.zip/workspace/extracted-artifacts/synthetics-recorder-0.0.1-beta.6-mac-arm64.zip.extracted/Elastic Synthetics Recorder.app/Contents/Frameworks/ReactiveObjC.framework/ReactiveObjC",
      "message": "The signature of the binary is invalid.",
      "docUrl": null,
      "architecture": "arm64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "synthetics-recorder-0.0.1-beta.6-mac-arm64.zip.zip/workspace/extracted-artifacts/synthetics-recorder-0.0.1-beta.6-mac-arm64.zip.extracted/Elastic Synthetics Recorder.app/Contents/Frameworks/Squirrel.framework/Squirrel",
      "message": "The signature of the binary is invalid.",
      "docUrl": null,
      "architecture": "arm64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "synthetics-recorder-0.0.1-beta.6-mac-arm64.zip.zip/workspace/extracted-artifacts/synthetics-recorder-0.0.1-beta.6-mac-arm64.zip.extracted/Elastic Synthetics Recorder.app/Contents/Frameworks/Mantle.framework/Mantle",
      "message": "The signature of the binary is invalid.",
      "docUrl": null,
      "architecture": "arm64"
    }
  ]
}
```

I have no idea why the file is called `synthetics-recorder-0.0.1-beta.6-mac-arm64.zip.zip` - but regardless, could you kindly check that your app is properly following [Apple guidelines](https://developer.apple.com/library/archive/technotes/tn2206/_index.html#//apple_ref/doc/uid/DTS40007919-CH1-TNTAG201) with regards to Nested Code.

<img width="2535" alt="image" src="https://user-images.githubusercontent.com/124953/191765990-c56b35c9-90e0-4137-8d75-f1273bd8ade9.png">

**Comment by kyungeunni** [comment]
A bit of update:

@nkammah and I caught up in Slack and verified that we can notarize the app that is signed by our CD with `xcrun notarytool submit` command. The next step will be creating a Jenkins job that signs, notarizes, and packages the Electron-based app.

In the meantime, we can notarize the latest release manually so that it is not blocked by the mentioned task above.

**Comment by v1v** [comment]
> The next step will be creating a Jenkins job that signs, notarizes, and packages the Electron-based app.

IIUC, that's already in place in the existing Release stage -> https://github.com/elastic/synthetics-recorder/blob/1559e6585d01978b43613215cf38783d9cc11370/.ci/Jenkinsfile#L148-L183, which uses:

* the CI worker -> `agent { label 'worker-c07yc0cejyvy' } `
* the function [runReleaseWithKeychain](https://github.com/elastic/synthetics-recorder/blob/1559e6585d01978b43613215cf38783d9cc11370/.ci/Jenkinsfile#L302-L315), which it is the function that was implemented in https://github.com/elastic/synthetics-recorder/pull/238 to sign the macos artifacts using the keychain stored in Vault.

If the notorizing is something to be done by another CI job provided by the Release team, then it could be similarly done as the one for signing the artifacts with the `gpg` in https://github.com/elastic/synthetics-recorder/blob/1559e6585d01978b43613215cf38783d9cc11370/.ci/Jenkinsfile#L221-L224, pretty much trigger another build from the  CI pipeline and wait for it.

Let me know if the above is not the way to do it, and how you envision the macos notarisation.
